[melody] Allow /useitem during melody

cord-b opened the pull request:
# Rules / Behavior
- `/useitem` will be allowing during melody.
- Before use, there is a 150ms delay (from the end of the song) until the click happens.
- After use, there is a 150ms delay before the Melody resumes.
- Activating an item manually (e.g. Right clicking UI or Hotbutton) still has no effect (to prevent client issues).

# Why
The reason for introducing this support is to get feature parity will old bard song twisting. Gaps in casting used to exist when twisting songs (between `/stopsong` and `/cast #`), and it was possible to use a clicky item in this gap if timed correctly. This was actually sort of trivial to perform if you bound your Song button and your item button to the same keybind, and just double tapped it.

Due to the complex nature of Melody and the Client casting state manipulation, this capability was lost due to its current implementation. With this new change, we can again have limited support for using clickies in this time interval between songs via the `/useitem` command.

# Implementation Notes
We took this approach in order to have better safety and preserve client integrity. By going through /useitem, this allows us to control safe timing gaps between songs and clickies. This avoids triggering spell timing vulnerabilities in the client.

This is also more balanced from a design side as well, by introducing a tradeoff/penalty of an additional 150ms extra delay when using a clicky item between melody songs. This wouldn't have been enforceable if players manually activated items with the native UI, and also would have been more error prone for issues/bugs.

By using the `/useitem` command, we can make sure the two code paths (useitem and Melody) are aware of each other and therefore avoid trying to both act at the same time. They can coordinate their timing to ensure the clicky is done safely each time, with predictable timing.

Directly right-clicking an item during a melody will be ignored as usual, to avoid unexpected behavior. This is handled automatically because the Melody will keep the client thinking it is in a casting/singing state at all times.

# Other Changes
- Moved `/useitem` implementation to `Zeal::EqGame::use_item(..)` so it can be shared between `/useitem` and Melody's use of it.

cord-b commented:
May need a stamp of approval from the powers at be, but wanted to give yall a chance to review. Briefly testing this it feels within the spirit of the game and not crossing any boundaries. 

For all the Breath of Harmony & SS BP enjoyers, those will finally be usable without a massive amount of clunkyness.

CoastalRedwood commented on Zeal/commands.cpp:
```diff
@@ -187,53 +187,26 @@ ChatCommands::ChatCommands(ZealService* zeal)
-				}
-				if (item_index < 21)
-					item = chr->InventoryItem[item_index];
+				if (char_info->Class == Zeal::EqEnums::ClassTypes::Bard && self->ActorInfo->CastingSpellId != kInvalidSpellId)
```
It's not clear to this will work properly with normal steady state (non-melody) singing (melody is off).

Let's change melody's use_item to return a boolean true if melody is active and the use_item was successfully queued (let melody's logic decide if it will handle it or not).  Something like below:

```
if (char_info->Class == Zeal::EqEnums::ClassTypes::Bard && ZealService::get_instance()->melody->use_item(item_index))
    return true;  // Melody is active and has queued the click for execution.

Zeal::EqGame::use_item(item_index);
```

cord-b replied:
Due to the non-Melody steady state and the timeout system, the user would see "You must stop casting to cast a spell!" when pressing /useitem and they were manually singing/casting something while Melody was off.

CoastalRedwood replied:
I wasn't sure this would work in the normal singing case since the melody() tick exits immediately when melody isn't active.  The use_item values would get ignored.

```
void Melody::tick()
{
    if (!songs.size())
        return;
```

CoastalRedwood commented on Zeal/melody.cpp:
```diff
@@ -97,6 +112,14 @@ void Melody::end()
+{
+    // Set fields so use_item(item_index) will execute during tick().
+    // tick() will handle checking if this was pressed at an allowed song gap timing.
+    use_item_index = item_index;
```
Let's change this to:
```
bool Melody::use_item(int index) {
     if (!songs.size() || index < 0)
        return false;
     use_item_index = item_index; 
     return true;
}
```

cord-b replied:
The index < 0 could cause a failed clicky if the user did something like this. So I'm thinking of an alternative way to handle that.

- Has a melody/song casting (e.g. casting_melody_spell_id = 123, index = 2)
- `start()`s a new Melody, but previous song is still finishing (e.g. casting_melody_spell_id = -1, index = -1)
- Played issues `/useitem` command to also queue up an item before the next song starts.

CoastalRedwood replied:
Sorry, maybe I'm a little slow this Sunday morning.  How would index < 0 cause a failed clicky? Melody bails out here if the index is < 0, so that can't fail.   The normal external use_item logic is executed.

But I did notice that it is probably worth resetting use_item_index = -1 before that return false above (clear any queued value).

cord-b replied:
Yeah I think this talk can get a bit abstract until the refactor starts to shape up. Anyway, I understand all the new goals/requirements you're getting at and will take those into account.

Sounds like for test cases we would have:
- /useitem then /melody -> the useitem is cleared
- /melody then /useitem -> the usetem is queued

CoastalRedwood commented on Zeal/melody.cpp:
```diff
@@ -158,6 +181,14 @@ void Melody::tick()
 
     ULONGLONG current_timestamp = GetTickCount64();
+
+    if (use_item_timeout && current_timestamp > use_item_timeout)
```
Let's ditch the timeout.

CoastalRedwood commented on Zeal/melody.cpp:
```diff
@@ -182,6 +213,16 @@ void Melody::tick()
     stop_current_cast();  //abort bard song if active.
 
+    // Execute a pending use_item() call here
+    if (use_item_timeout >= current_timestamp)
```
Let's rely on use_item_index >= 0
```
if (use_item_index >= 0) {
    bool success = Zeal::EqGame::use_item(use_item_index);
    use_item_index = -1;
    if (success) {
        casting_visible_timestamp = current_timestamp; // Pushes back the start of next song by MELODY_SONG_INTERVAL ms.
        return;
    }
}
```

CoastalRedwood commented on Zeal/melody.h:
```diff
@@ -9,6 +9,7 @@ class Melody
 	bool start(const std::vector<int>& new_songs); //returns true if no errors
 	void end();
 	void handle_stop_cast_callback(BYTE reason, WORD spell_id);
+	void use_item(int item_index); // asks Melody to perform /useitem during the current song gap.
```
bool return

CoastalRedwood commented on Zeal/melody.h:
```diff
@@ -18,5 +19,6 @@ class Melody
-};
-
+	int use_item_index = -1; // The pending use_item() to try.
+	ULONGLONG use_item_timeout = 0; // The max timestamp until the pending use_item() gives up. Cleared when use_item() runs.
```
remove timeout

CoastalRedwood commented on Zeal/melody.h:
```diff
@@ -18,5 +19,6 @@ class Melody
 	WORD casting_melody_spell_id = kInvalidSpellId; // Current melody song being cast. Is only a valid id while cast window is visible (actively casting).
-};
-
+	int use_item_index = -1; // The pending use_item() to try.
```
Make sure to reset use_item_index to -1 any time melody is reset (look for anywhere current_index = -1).

cord-b replied:
If a player has an active Melody and switches to a new Melody while having a queued clicky, my take on that is we'd still want it to fire between those song casts, as the only difference is the player changed the song rotation, but still wants to have a melody playing.

This was one of the benefits of the timeout system as it could handle this naturally without having to react to the current_index being manipulated or reset.

CoastalRedwood replied:
I disagree on this and actually would argue it's one of the disadvantages.  I'd vote that starting a new melody is a higher priority than a queued clicky (especially something like a 6 second clicky cast).

CoastalRedwood replied:
Remember that it should be easy to just add the clicky again after switching the melody up.

CoastalRedwood reviewed (commented):
I chatted with Salty.  We don't think there will be major objections as long as the clicky isn't automated directly into the melody list.  So let's just simplify this to let use_item be queued and not worry about the timing (drop all the time tracking and comments).

I made a list of suggested small changes (probably incomplete) to support this.

Does that make sense

CoastalRedwood commented:
> I think it could still be beneficial to have some form of guard against a really stale `/useitem` coming through by the time the player forgot about it.
> 
> I'm not against generally refactoring this a bit as suggested, but there could be some edge cases that feel weird with unlimited queueing.
> 
> One example being if you were spamming your `/useitem <SingingSteelBP>`, you might accidentally queue it again as soon as it starts casting. Back to back casting is supported, but I'm not sure if a ~6 second stale click is something the person would intend to have it repeat the cast for. I think a general player expectation is to press the next /useitem relatively close to when the cast is ending, and the rest would not queue it up.
> 
> I think maybe a 1-2 second window would be fine as balance, which is plenty to make it low-effort but doesn't get to the point where the queueing starts taking away player control of actions in the future if they need to react to something and change their plans.
> 
> Or we could extend `/useitem <slot> <timeout>` as an optiontal 2nd parameter for their own desired timeout. Though I assume regular players wouldn't use advanced options like that.

Okay, I don't have a problem if you want to keep a stale timeout check for that SS BP type long-cast corner case.  I'd consider just making it 3.5 seconds though so they could hit it anytime within a normal melody and it will be 'guaranteed' to happen.  If the SS BP has been going for 3 seconds and you hit it again, well, then probably it should get triggered to repeat.  And I would make that timeout a secondary check in that if useitem is >= 0 branch.

Let's make sure that melody start/end clear / resets all states.

cord-b commented:
Sounds good I will tinker on this for a bit and see how it starts to feel.

cord-b commented:
Been kinda busy but plan to tackle this in the next few days.

CoastalRedwood commented:
> Been kinda busy but plan to tackle this in the next few days.

Sounds good. I just got a SS BP, and it would be a nice QoL improvement to have this versus doing the current /stopsong, /useitem 16 macro and then restarting my melody.  Especially when things get a bit laggy.

If you want me to take a swing at it, just let me know.  Finally finished most of the map stuff.

iamclint commented:
Can you message me on discord cord-b

cord-b commented:
Should be good to go now:
- Rebased on latest master, made sure to port over the newest version of the use_item() implementation before moving it to `Zeal::EqFunctions::use_item()` (supports /useitem while sitting)
- `tick()` looks for `use_item_index >= 0` now.
- Removed the verbose timeout-related comments/fields. A timeout of 3.5 seconds is used as a final/secondary check, but is mostly simplified.
- `use_item_index` is reset on melody starting/stopping.

CoastalRedwood reviewed (commented):
I took a quick look and it looked pretty good (nice and simple).  I'll give it a quick test run tomorrow morning to look for any non-intuitive corner cases.

CoastalRedwood commented on Zeal/melody.cpp:
```diff
@@ -93,10 +95,22 @@ void Melody::end()
 
+bool Melody::use_item(int item_index)
+{
+    if (songs.empty())
```
Possibly change to i`f (songs.empty() || (item_index < 0))` to avoid silent failures (ignoring it).

CoastalRedwood reviewed (approved):
A morning hike delayed my testing, but I finally had a chance.  It generally worked great and is a nice QoL addition simplifying things (but still requiring manual action).  /useitem worked while sitting, standing, with and without melody, etc.

Issues I discovered are below.  It would be nice to fix (2) somehow, but I don't think it is required since if you are using hotkeys, the /stopsong gets blocked anyways due to issue (1).   Issues (1) and (3) are future nice to fixes.  (4) would be a quick fix

(1) As you mentioned, the hotkeys work differently than the action keys or command line.  Not the problem of this PR, but it would be nice if to not see that "You can't use that command right now.." for a hotkey during a SS BP clicky cast while it allows queuing the action by pushing the social actions button directly.  If melody isn't playing, the Actions bar says "You must stop casting to cast this spell".

(2) There is an issue with /stopsong and clicky casts.  Maybe this is an existing melody.cpp issue with handling stopsong?  I don't recall noticing this previously, but I don't think this PR caused it.  If you start a SS BP 6 second cast (with or w/out melody active) and then do a stopsong (either command line directly or as part of a macro), the Invigorate casting bar goes away but it continues in the background, and then if you try to start a melody it flashes through a bunch of song ends / retries and the melody aborts.  If you just start a SS BP clicky cast and then do a /melody without a /stopsong, the invigorate stays up and it starts up normally.

(3) It would be nice if the /useitem behavior to queue during melody also worked for right clicks on inventory items.  I don't know the right click path and if there is a detour / hook for that.

(4) doing a /useitem -1 (or any negative) silently fails when melody is running.  See code comment.

cord-b commented:
> (2) There is an issue with /stopsong and clicky casts. 

Yep confirmed this also happens without Zeal installed at all, so it's a client/server incompatibility that's baked into the client. The Invigorate cast bar disappears and manually click a spell gem just briefly flashes / tries to start the song, but the server keeps interrupting you, and eventually Invigorate casts.

I can fix this by making melody hook /stopsong first and we have two options/behaviors that can trigger when used by a bard:
- (2a) "Ignore" - If the spell in progress isn't a memorized spell gem, short-circuit the `/stopsong` into a no-op so the client doesn't try to stop it.
- (2b) "Force Interrupt" - Make the `/stopsong` actually stop the cast: call `stop_current_cast()` and `end()`, rather than falling through to the bugged `/stopsong` implementation.


> (3) It would be nice if the /useitem behavior to queue during melody also worked for right clicks on inventory items. I don't know the right click path and if there is a detour / hook for that.

Yeah it's out of my wheelhouse to figure that out, but if anyone finds a way to hook that cast attempt or even hotbar press, we could likely route that to melody and process it safely.

> (4) doing a /useitem -1 (or any negative) silently fails when melody is running. See code comment.

Sounds good, simple fix. Thanks for catching that.

CoastalRedwood commented:
I like your 2(a) suggestion if you have time to try it out.  Other classes don't have a /stopcast, so it's similar behavior to ignore the /stopsong on a clicky.  And then my usual melody macros that almost always start with /stopsong will work to not interrupt the long clicky (just queues up I think).

With or without this PR look good to me.

CoastalRedwood reviewed (approved):
Ran a few quick tests.  Issue (4) and (2) are fixed with the second commit.
